商品購入機能

fatgranddad opened the pull request:
## What
1. 必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができること。
2. 商品出品時に登録した情報のうち、「商品名・商品画像・価格・配送料の負担」の4つの情報が、見本アプリと同様の形で表示されること。
3. ログイン状態の場合は、自身が出品していない販売中商品の商品購入ページに遷移できること。
4. ログイン状態の場合でも、自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること。
5. ログイン状態の場合でも、自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移すること。
6. ログアウト状態の場合は、商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること。
7. 購入が完了したら、トップページに遷移すること。
8. エラーハンドリングができること（入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示されること）。
9. エラーハンドリングによって購入ページに戻った場合でも、入力済みの項目（カード番号・有効期限・セキュリティコード以外）は消えないこと。
10. エラーハンドリングの際、重複したエラーメッセージが表示されないこと。

## Why
1. 利用者がスムーズに商品を購入できるようにするため。
2. 利用者が必要な情報を一目で確認できるようにし、購入意志の決定を助けるため。
3. 利用者が興味を持った商品に対してスムーズにアクセスできるようにするため。
4. 売却済み商品に対して不正な購入を防ぐため。
5. 出品者自身が自分の商品を購入するという不正を防ぐため。
6. 未ログインユーザーの購入操作を防ぎ、セキュリティを確保するため。
7. 購入後の処理を明確にし、ユーザーが購入完了を認識できるようにするため。
8. ユーザーが入力ミスを即座に認識でき、修正できるようにするため。
9. ユーザーがエラーで戻された場合でも、再入力の手間を減らすため。
10. ユーザーが同じエラーメッセージに繰り返し遭遇することでの混乱を避けるため。

## 検証動画
1. [必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる](https://gyazo.com/a08f267ee262a3b0f7a96b3e141c5ee1)
2. [入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される](https://gyazo.com/3b5fd817a82affe9e8649c6eedfd0fdc)
3. [ログイン状態でも、URLを直接入力して自身が出品していない売却済み商品の購入ページへ遷移しようとすると、トップページに遷移する](https://gyazo.com/9af523998b61dca09a006708b59d0629)
4. [ログイン状態で、URLを直接入力して自身が出品した商品の購入ページに遷移しようとすると、トップページに遷移する](https://gyazo.com/b520193c351a3070b43d4ebe26f6d0ab)
5. [ログアウト状態で、URLを直接入力して商品購入ページに遷移しようとすると、ログインページに遷移する](https://gyazo.com/d11328d1eceecb7d2ecbe6fe3060cbc5)
6. [売却済みの商品は、画像上に「sold out」の文字が表示される](https://gyazo.com/10b6953c901ee7e6629fb247811c80d4)
7. [ログイン状態でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない](https://gyazo.com/5131f2ded1f35e79f513883c6c6c8339)
8. [ログイン状態で、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する](https://gyazo.com/64b0ff812eda427d40871bfb299c1bde)
9. [テスト結果の画像](https://gyazo.com/946d77e11cb24cd87c386576f6c22652)

fatgranddad committed:
api設定_鍵設定前

fatgranddad committed:
orderとaddressモデルの追加

fatgranddad committed:
ordersビューを整理

fatgranddad committed:
環境変数の設定

fatgranddad committed:
sold outと表示機能追加

fatgranddad committed:
sold out表示_商品詳細機能

fatgranddad committed:
売却済みの商品には編集等ボタンを表示させない

fatgranddad committed:
詳細ページにて常にsold outと表示されてしまう事象を解消

fatgranddad committed:
売却済み商品は編集ページにアクセルするとトップページへ遷移する

fatgranddad committed:
自分の商品は購入できない機能

ログイン状態の場合でも、自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する

fatgranddad committed:
ログアウトの状態では購入できない機能

ログアウト状態の場合は、商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する

fatgranddad committed:
Rubocopを実行

TakashiKozuma commented on db/migrate/20240413084343_add_sold_to_items.rb:
```diff
@@ -0,0 +1,5 @@
+class AddSoldToItems < ActiveRecord::Migration[7.0]
+  def change
+    add_column :items, :sold, :boolean, default: false, null: false
+  end
+end
```
```Rb
class AddSoldToItems < ActiveRecord::Migration[7.0]
  def change
    add_column :items, :sold, :boolean, default: false, null: false
  end
end
```
READMEに存在しないカラムなので、ファイルごと削除しましょう。
売り切れかどうかの実装はカラムを増やすことなく、別の方法で実装が可能です。

fatgranddad replied:
修正しました。

TakashiKozuma commented on app/controllers/items_controller.rb:
```diff
@@ -53,4 +54,8 @@ def item_params
   end
+
+  def redirect_if_sold
+    redirect_to root_path, alert: '売却済みの商品は編集または削除できません。' if @item.sold
```
```Rb
redirect_to root_path, alert: '売却済みの商品は編集または削除できません。' if @item.sold
```
売り切れかどうかの判断はカラムを追加するのではなく、アソシエーションを利用して実装しましょう。

fatgranddad replied:
修正しました。

TakashiKozuma commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,71 @@
+  def update_item_sold_status
+    @item.update!(sold: true)
+  end
+end
```
コード全般に関して、商品に新たなカラムを追加するのではなく、アソシエーションを利用して購入済みかどうかの判断を行うように修正しましょう。

fatgranddad replied:
修正しました。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
コメントを致しましたので、ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

fatgranddad committed:
アソシエーションを利用して売却済みを判断

soldカラムを追加していたため、該当カラムの削除、また該当カラムを参照していたコードを削除。アソシエーションを利用した売却判断ができるように修正した。

fatgranddad committed:
Rubocopを実行

fatgranddad commented:
追加してしまったsoldカラムの削除、また該当カラムを参照していたコードを削除。
アソシエーションを利用した売却判断ができるように修正しました。
再度、コードレビューのほどお願いいたします。

Akinori0123 commented on app/models/purchase_form.rb:
```diff
@@ -0,0 +1,33 @@
+  validates :city, presence: true
+  validates :street_address, presence: true
+  validates :phone_number, presence: true, format: { with: /\A\d{10,11}\z/ }  # 電話番号は、「09012345678」または「0312345678」の形式に一致する必要があります（10桁または11桁の数字）。
+  validates :card_token, presence: true
```
`presense: true`のように複数回使用するバリデーションについては`with_options`を使用するとまとめて記載することが可能です。

以下カリキュラムの`with_options`の解説をご参考ください。
```
最終課題補足カリキュラム
1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう
```

fatgranddad replied:
修正しました。

Akinori0123 commented on app/models/order.rb:
```diff
@@ -0,0 +1,4 @@
+class Order < ApplicationRecord
+  belongs_to :user
+  belongs_to :item
+end
```
①
OrderモデルファイルにAddressとのアソシエーションを追加しましょう。

②
※該当ファイルにコメントができないためこちらに失礼します。
もしUserモデルに対してOrderとのアソシエーションが記載されていない場合は、Userモデルのアソシエーションも修正しましょう。

fatgranddad replied:
追加しました。

Akinori0123 commented on spec/models/purchase_form_spec.rb:
```diff
@@ -0,0 +1,79 @@
+        @purchase_form.card_token = ''
+        @purchase_form.valid?
+        expect(@purchase_form.errors.full_messages).to include("Card token can't be blank")
+      end
```
**◎異常系テスト**
要件に沿って以下のテストも実装しましょう。
・購入者（user_id）が空だと購入できない
・商品（item_id）が空だと購入できない
・電話番号に半角数値以外の文字が含まれると購入できない

fatgranddad replied:
追加しました。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
前回の修正確認いたしました！アソシエーションによる売却済み判断の記述良いですね。
追加でコメントいたしましたのでご確認のほどよろしくお願いいたします。

fatgranddad committed:
presence: trueをまとめる

with_options presence: true do ~ endでまとめた

fatgranddad committed:
不足分のアソシエーションを追加

fatgranddad committed:
テストコードを追加

追加テストコードが失敗し、バリデーションの不足を発見したため、
PurchaseForm モデルに、user_id と item_idのバリデーション（presence: true）を設定。

fatgranddad commented:
レビューありがとうございます。ご指摘いただきました以下を対応いたしました。
- バリデーションをまとめる
- アソシエーションの追加
- テストコードの追加（テスト失敗にともなう、不足分のバリデーションを追加）し、成功を確認しています。

再度のレビューをお願いします。

検証画像
[テストコードの結果](https://gyazo.com/79d6759e28f8266e47a65610651d4d03)

yuri240 reviewed (commented):
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

ここまでの実装、大変お疲れ様でした。
引き続き、応援しております。

fatgranddad committed:
Rubocopを実行
